// jnitrace-engine/dist/utils/java_method.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("./types");
const SEMI_COLON_OFFSET = 1;
/**
 * Abstracts a Java method referenced in native code.
 */
class JavaMethod {
    constructor(signature) {
        const primitiveTypes = ["B", "S", "I", "J", "F", "D", "C", "Z", "V"];
        let isArray = false;
        let isRet = false;
        const jParamTypes = [];
        let jRetType = "unknown";
        for (var i = 0; i < signature.length; i++) {
            if (signature.charAt(i) === "(") {
                continue;
            }
            if (signature.charAt(i) === ")") {
                isRet = true;
                continue;
            }
            if (signature.charAt(i) === "[") {
                isArray = true;
                continue;
            }
            let jtype = "unknown";
            if (primitiveTypes.includes(signature.charAt(i))) {
                jtype = signature.charAt(i);
            }
            else if (signature.charAt(i) === "L") {
                var end = signature.indexOf(";", i) + SEMI_COLON_OFFSET;
                jtype = signature.substring(i, end);
                i = end - SEMI_COLON_OFFSET;
            }
            //TODO DELETE
            if (isArray) {
                jtype = "[" + jtype;
            }
            if (!isRet) {
                jParamTypes.push(jtype);
            }
            else {
                jRetType = jtype;
            }
            isArray = false;
        }
        this._ = signature;
        this._params = jParamTypes;
        this._ret = jRetType;
    }
    /**
     * Get the Java param types for the method.
     */
    get params() {
        return this._params;
    }
    /**
     * Get the Java param types as native jtypes.
     */
    get nativeParams() {
        const nativeParams = [];
        this._params.forEach((p) => {
            const nativeJType = types_1.Types.convertJTypeToNativeJType(p);
            nativeParams.push(nativeJType);
        });
        return nativeParams;
    }
    /**
     * Get the Java params as Frida native types.
     */
    get fridaParams() {
        const fridaParams = [];
        this._params.forEach((p) => {
            const nativeJType = types_1.Types.convertJTypeToNativeJType(p);
            const fridaType = types_1.Types.convertNativeJTypeToFridaType(nativeJType);
            fridaParams.push(fridaType);
        });
        return fridaParams;
    }
    /**
     * Get the Java return type of the method.
     */
    get ret() {
        return this._ret;
    }
    /**
     * Get the Java return type as a Frida native type.
     */
    get fridaRet() {
        const jTypeRet = types_1.Types.convertJTypeToNativeJType(this._ret);
        return types_1.Types.convertNativeJTypeToFridaType(jTypeRet);
    }
}
exports.JavaMethod = JavaMethod;
