// jnitrace-engine/dist/jni/jni_thread_manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class JNIThreadManager {
    constructor() {
        this.threads = {};
        this.shadowJavaVM = NULL;
    }
    getJavaVM() {
        return this.shadowJavaVM;
    }
    hasJavaVM() {
        return !this.shadowJavaVM.isNull();
    }
    setJavaVM(javaVM) {
        this.shadowJavaVM = javaVM;
    }
    getJNIEnv(threadId) {
        if (this.threads[threadId] !== undefined) {
            return this.threads[threadId];
        }
        else {
            return NULL;
        }
    }
    hasJNIEnv(threadId) {
        return !this.getJNIEnv(threadId).isNull();
    }
    setJNIEnv(threadId, jniEnv) {
        this.createEntry(threadId, jniEnv);
    }
    needsJNIEnvUpdate(threadId, jniEnv) {
        const entry = this.getEntry(threadId);
        if (entry === undefined || !entry.equals(jniEnv)) {
            return true;
        }
        return false;
    }
    createEntry(threadId, jniEnv) {
        this.threads[threadId] = jniEnv;
    }
    getEntry(threadId) {
        return this.threads[threadId];
    }
}
exports.JNIThreadManager = JNIThreadManager;
